server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4181,10 +4183,25 @@ public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long
         waitForPrimaryTermAndGeneration(getOperationPrimaryTerm(), segmentGeneration, listener);
     }
 
+    private final SubscribableListener<Void> waitForPrimaryTermAndGenerationListeners = new SubscribableListener<Void>();
+
+    private void checkAndCallWaitForPrimaryTermAndGenerationListeners() {
+        if (state == IndexShardState.STARTED || state == IndexShardState.CLOSED) {
+            waitForPrimaryTermAndGenerationListeners.onResponse(null);
+        }
```

Comment:
I am not familiar with the code, so this question is for my own learning: Since we complete the listener once the state is either STARTED or CLOSED, my understanding is that it implies the state cannot change from these two states to other states that are _not_ either of the two. In another word, there is no need to revoke the readiness of the SubscribableListener. Am I getting this right?

Reply:
I believe that recovery is a chain that starts from CREATED state and then goes to RECOVERING, POST_RECOVERY, STARTED, and CLOSED (when the shard is closed or it fails during the whole process). After being CLOSED, the shard can be re-allocated, but that should go again through the same chain of states (recovery can be basically a no-op if for example the files are already there on the same machine). If my belief is wrong, I hope the team can correct me :)